File: backend/app/schemas/feasibility.py
"""Pydantic models for feasibility rule and assessment APIs."""

from __future__ import annotations

from typing import Literal

from app.schemas.finance import AssetFinancialSummarySchema
from pydantic import BaseModel, Field

FeasibilityRuleSeverity = Literal["critical", "important", "informational"]
FeasibilityRuleStatus = Literal["pass", "fail", "warning"]


class BuildEnvelopeSnapshot(BaseModel):
    """Optional snapshot of zoning envelope inputs."""

    site_area_sqm: float | None = Field(default=None, gt=0)
    allowable_plot_ratio: float | None = Field(default=None, gt=0)
    max_buildable_gfa_sqm: float | None = Field(default=None, gt=0)
    current_gfa_sqm: float | None = Field(default=None, ge=0)
    additional_potential_gfa_sqm: float | None = Field(default=None)


class NewFeasibilityProjectInput(BaseModel):
    """Input details describing the project under assessment."""

    name: str = Field(..., min_length=1)
    site_address: str = Field(..., min_length=1)
    site_area_sqm: float = Field(..., gt=0)
    land_use: str = Field(..., min_length=1)
    target_gross_floor_area_sqm: float | None = Field(None, gt=0)
    building_height_meters: float | None = Field(None, gt=0)
    build_envelope: BuildEnvelopeSnapshot | None = None


class FeasibilityRule(BaseModel):
    """Rule definition returned to the frontend wizard."""

    id: str
    title: str
    description: str
    authority: str
    topic: str
    parameter_key: str
    operator: str
    value: str
    unit: str | None = None
    severity: FeasibilityRuleSeverity
    default_selected: bool = False


class FeasibilityRulesSummary(BaseModel):
    """Summary describing how the recommended rules were chosen."""

    compliance_focus: str
    notes: str | None = None


class FeasibilityRulesResponse(BaseModel):
    """Response payload for the feasibility rules endpoint."""

    project_id: str
    rules: list[FeasibilityRule]
    recommended_rule_ids: list[str]
    summary: FeasibilityRulesSummary


class FeasibilityAssessmentRequest(BaseModel):
    """Request payload for evaluating the selected rules."""

    project: NewFeasibilityProjectInput
    selected_rule_ids: list[str]


class RuleAssessmentResult(FeasibilityRule):
    """Assessment outcome for a rule."""

    status: FeasibilityRuleStatus
    actual_value: str | None = None
    notes: str | None = None


class BuildableAreaSummary(BaseModel):
    """High-level metrics produced by the feasibility engine."""

    max_permissible_gfa_sqm: int
    estimated_achievable_gfa_sqm: int
    estimated_unit_count: int
    site_coverage_percent: float
    remarks: str | None = None


class AssetOptimizationRecommendation(BaseModel):
    """Asset-specific optimisation recommendation surfaced to clients."""

    asset_type: str
    allocation_pct: float
    allocated_gfa_sqm: int | None = None
    nia_efficiency: float | None = None
    target_floor_height_m: float | None = None
    parking_ratio_per_1000sqm: float | None = None
    rent_psm_month: float | None = None
    stabilised_vacancy_pct: float | None = None
    opex_pct_of_rent: float | None = None
    estimated_revenue_sgd: float | None = None
    estimated_capex_sgd: float | None = None
    fitout_cost_psm: float | None = None
    absorption_months: int | None = None
    risk_level: str | None = None
    heritage_premium_pct: float | None = None
    notes: list[str] = Field(default_factory=list)


class FeasibilityAssessmentResponse(BaseModel):
    """Response payload produced after evaluating the project."""

    project_id: str
    summary: BuildableAreaSummary
    rules: list[RuleAssessmentResult]
    recommendations: list[str]
    asset_optimizations: list[AssetOptimizationRecommendation]
    asset_mix_summary: AssetFinancialSummarySchema | None = None


__all__ = [
    "BuildEnvelopeSnapshot",
    "BuildableAreaSummary",
    "FeasibilityAssessmentRequest",
    "FeasibilityAssessmentResponse",
    "FeasibilityRule",
    "FeasibilityRuleSeverity",
    "FeasibilityRuleStatus",
    "FeasibilityRulesResponse",
    "FeasibilityRulesSummary",
    "AssetOptimizationRecommendation",
    "AssetFinancialSummarySchema",
    "NewFeasibilityProjectInput",
    "RuleAssessmentResult",
]

File: frontend/src/api/feasibility.ts
import { buildUrl } from './shared'

const API_PREFIX = 'api/v1/feasibility'

export interface BuildEnvelopeSnapshot {
  siteAreaSqm?: number | null
  allowablePlotRatio?: number | null
  maxBuildableGfaSqm?: number | null
  currentGfaSqm?: number | null
  additionalPotentialGfaSqm?: number | null
}

export interface NewFeasibilityProjectInput {
  name: string
  siteAddress: string
  siteAreaSqm: number
  landUse: string
  targetGrossFloorAreaSqm?: number | null
  buildingHeightMeters?: number | null
  buildEnvelope?: BuildEnvelopeSnapshot | null
  typFloorToFloorM?: number | null
  efficiencyRatio?: number | null
  structureType?: string | null
  mepLoadWpsm?: number | null
}

export type FeasibilityRuleSeverity = 'critical' | 'important' | 'informational'
export type FeasibilityRuleStatus = 'pass' | 'fail' | 'warning'

export interface FeasibilityRule {
  id: string
  title: string
  description: string
  authority: string
  topic: string
  parameterKey: string
  operator: string
  value: string
  unit?: string | null
  severity: FeasibilityRuleSeverity
  defaultSelected: boolean
}

export interface FeasibilityRulesSummary {
  complianceFocus: string
  notes?: string | null
}

export interface FeasibilityRulesResponse {
  projectId: string
  rules: FeasibilityRule[]
  recommendedRuleIds: string[]
  summary: FeasibilityRulesSummary
}

export interface FeasibilityAssessmentRequest {
  project: NewFeasibilityProjectInput
  selectedRuleIds: string[]
}

export interface RuleAssessmentResult extends FeasibilityRule {
  status: FeasibilityRuleStatus
  actualValue?: string | null
  notes?: string | null
}

export interface BuildableAreaSummary {
  maxPermissibleGfaSqm: number
  estimatedAchievableGfaSqm: number
  estimatedUnitCount: number
  siteCoveragePercent: number
  remarks?: string | null
}

export interface AssetConstraintViolation {
  constraintType: string
  severity: string
  message: string
  assetType?: string | null
}

export interface AssetFinancialSummarySchema {
  totalEstimatedRevenueSgd?: number | null
  totalEstimatedCapexSgd?: number | null
  dominantRiskProfile?: string | null
  notes?: string[]
}

export interface AssetOptimizationRecommendation {
  assetType: string
  allocationPct: number
  allocatedGfaSqm?: number | null
  niaEfficiency?: number | null
  targetFloorHeightM?: number | null
  parkingRatioPer1000Sqm?: number | null
  rentPsmMonth?: number | null
  stabilisedVacancyPct?: number | null
  opexPctOfRent?: number | null
  estimatedRevenueSgd?: number | null
  estimatedCapexSgd?: number | null
  fitoutCostPsm?: number | null
  absorptionMonths?: number | null
  riskLevel?: string | null
  heritagePremiumPct?: number | null
  notes: string[]
  niaSqm?: number | null
  estimatedHeightM?: number | null
  totalParkingBaysRequired?: number | null
  revenueBasis?: string | null
  constraintViolations: AssetConstraintViolation[]
  confidenceScore?: number | null
  alternativeScenarios: string[]
}

export interface FeasibilityAssessmentResponse {
  projectId: string
  summary: BuildableAreaSummary
  rules: RuleAssessmentResult[]
  recommendations: string[]
  assetOptimizations: AssetOptimizationRecommendation[]
  assetMixSummary?: AssetFinancialSummarySchema | null
  constraintLog: AssetConstraintViolation[]
  optimizerConfidence?: number | null
}

export async function runFeasibilityAssessment(
  request: FeasibilityAssessmentRequest,
  signal?: AbortSignal,
): Promise<FeasibilityAssessmentResponse> {
  const response = await fetch(buildUrl(`${API_PREFIX}/assessment`), {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(request),
    signal,
  })

  if (!response.ok) {
    const detail = await response.text()
    throw new Error(
      detail || `Request to ${API_PREFIX}/assessment failed with ${response.status}`,
    )
  }

  return (await response.json()) as FeasibilityAssessmentResponse
}

export interface EngineeringDefaults {
  typFloorToFloorM: number
  efficiencyRatio: number
}

export async function getEngineeringDefaults(
  signal?: AbortSignal,
): Promise<EngineeringDefaults> {
  const response = await fetch(buildUrl(`${API_PREFIX}/defaults`), {
    method: 'GET',
    headers: { 'content-type': 'application/json' },
    signal,
  })

  if (!response.ok) {
    const detail = await response.text()
    throw new Error(
      detail || `Request to ${API_PREFIX}/defaults failed with ${response.status}`,
    )
  }

  return (await response.json()) as EngineeringDefaults
}

// Alias for backward compatibility
export { runFeasibilityAssessment as submitFeasibilityAssessment }
